### PR TITLE
Add compute lab

### PR DESCRIPTION
**Pull request by teodutu**

### CODE REVIEW

File: content/chapters/compute/lab/README.md
Comment (razvand) [inline content/chapters/compute/lab/README.md:149]:
```suggestion
This is **not** a syscall, but a wrapper over the most common thread-management API in POSIX-based operating systems (such as Linux, FreeBSD, macOS): POSIX Threads or `pthreads`.
```

File: content/chapters/compute/lab/README.md
Comment (razvand) [inline content/chapters/compute/lab/README.md:194]:
```suggestion
Relevant flags passed as arguments when creating threads are documented in [`clone`'s man page](https://man.archlinux.org/man/clone3.2.en):
```

File: content/chapters/compute/lab/README.md
Comment (razvand) [inline content/chapters/compute/lab/README.md:197]: This is a rather out-of-nowhere reference to TLS. I assume it will be detailed sometime in the lab text.

File: content/chapters/compute/lab/README.md
Comment (razvand) [inline content/chapters/compute/lab/README.md:201]:
```suggestion
When creating a process, `clone` creates this new thread within a new separate address space.
The creation of a new address space and a new thread equates to the creation of a new process.
```

**Review by razvand — CHANGES_REQUESTED** [review]
Thanks, @teodutu . This is a good start.

I think at times you go into much depth and on not-so-relevant tracks. I would keep a "main, simpler" narrative and leave the rest, such as D threading model, benchmarking internals as "extra" / bonus topics.

See my comments below.

**Review by Alex-deVis — COMMENTED** [review]
We could develop the section about creating processes by providing a `fork` + `exec` example. It would also give better insight for examples with strace where `execve` is featured.

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:6]: It explains what is being measured as part of a benchmark to get a better idea about what a CPU does adn why there is a need for an OS. All in all, it's less than 15 lines and then moves on introduce the OS. I don't think it's so elaborate, especially since the links aren't menat to be analysed in detail, but merely skimmed over.

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:197]: Yes. I think this is alright since this section comes at the end of the "regular" lab.

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:141]: I said above that the difference should be small. Should I remove this paragraph or rephrase it?

File: content/chapters/compute/lab/support/sum-array/d/generate_random_array.d
Comment (teodutu) [reply content/chapters/compute/lab/support/sum-array/d/generate_random_array.d:3]: To better separate differents "parts" of the code, but now that you asked, they're pretty pointless. I removed the extra line.

File: content/chapters/compute/lab/quiz/processes-speedup.md
Comment (teodutu) [reply content/chapters/compute/lab/quiz/processes-speedup.md:18]: I meant to send the message that there will always be parts of an app that cannot be paralellised and, thus, the speedup will never be equal to the number of processes. But you're right that it's ambiguous. It should be clearer now.

**Comment by teodutu** [comment]
@Alex-deVis 
> We could develop the section about creating processes by providing a fork + exec example

I deliberately tried to avoid this. `fork` + `exec` is too complex to be placed at the beginning of the lab. It will be covered, but towards the end of the lab or maybe in the `Arena`, because it's an implementation detail on POSIX systems.

File: content/chapters/compute/lab/README.md
Comment (razvand) [reply content/chapters/compute/lab/README.md:6]: OK, you can keep it. My concern is that it focuses on benchmarking and results which may be too much for students at this point to grasp.

File: content/chapters/compute/lab/README.md
Comment (razvand) [reply content/chapters/compute/lab/README.md:193]: I see. I would remove the blank line between them. Also, I would rephrase it to "This slight time difference is caused by process creation actions, which are costlier than thread creation actions." Or something in that note.

File: content/chapters/compute/lab/README.md
Comment (molecula2788) [inline content/chapters/compute/lab/README.md:679]: It should be mentioned that python threads are still useful when you have multiple io-bound tasks. Otherwise one would believe that they are completely useless.

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:101]: No, it was only a duplicate line

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:230]: It is a problem. I moved it after `Practice: Wait for Me!`.

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:369]: It is present in the [lecture PR](https://github.com/systems-cs-pub-ro/operating-systems-oer/pull/5). Once that is merged, it should render properly. There's no point in duplicating it.

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:310]: Right. I only left a forward reference to copy-on-write to make students curious.

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:567]: The scheduler itself also creates some additional threads, the output from `ltrace` is messy at best. I've just linked the implementation of `spawn`. It should be clearer.

File: content/chapters/compute/lab/solution/mini-shell/mini_shell.c
Comment (molecula2788) [inline content/chapters/compute/lab/solution/mini-shell/mini_shell.c:9]:
```suggestion
#include "utils.h"
```

File: content/chapters/compute/lab/solution/mini-shell/mini_shell.c
Comment (molecula2788) [inline content/chapters/compute/lab/solution/mini-shell/mini_shell.c:27]:
```suggestion
```
```
mini_shell.c:27:14: error: ‘var’ defined but not used [-Werror=unused-variable]
   27 | static char *var;
      |              ^~~
mini_shell.c:26:14: error: ‘value’ defined but not used [-Werror=unused-variable]
   26 | static char *value;
      |              ^~~~~
```

File: content/chapters/compute/lab/solution/mini-shell/mini_shell.c
Comment (molecula2788) [inline content/chapters/compute/lab/solution/mini-shell/mini_shell.c:95]:
```suggestion
		DIE(1, "fork");
		break;
```

File: content/chapters/compute/lab/solution/mini-shell/mini_shell.c
Comment (molecula2788) [inline content/chapters/compute/lab/solution/mini-shell/mini_shell.c:100]:
```suggestion
		DIE(1, "execvp");
		break;
```

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:145]: The point of this exercise is to teach students how to use `Popen()` so they can do so themselves in the next exercise.

File: content/chapters/compute/lab/solution/race-condition/asm/race_condition_lock.S
Comment (molecula2788) [inline content/chapters/compute/lab/solution/race-condition/asm/race_condition_lock.S:57]:
```suggestion
```

File: content/chapters/compute/lab/support/race-condition/asm/race_condition_lock.S
Comment (molecula2788) [inline content/chapters/compute/lab/support/race-condition/asm/race_condition_lock.S:57]:
```suggestion
```

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:145]: You're right. I've added a few phrases regarding this in the lab text.

File: content/chapters/compute/lab/solution/sleepy/Makefile
Comment (molecula2788) [inline content/chapters/compute/lab/solution/sleepy/Makefile:5]: I think this is a leftover from previous versions?
```
$ make
Makefile:4: *** target file 'clean' has both : and :: entries.  Stop.
```
Same in `support/sleepy/Makefile`.

File: content/chapters/compute/lab/README.md
Comment (molecula2788) [inline content/chapters/compute/lab/README.md:905]: I'm not sure if this is correct? If a user-level thread does a blocking io call, none of the other user-level threads can run.

File: content/chapters/compute/lab/quiz/tcb-libult-unikraft.md
Comment (teodutu) [reply content/chapters/compute/lab/quiz/tcb-libult-unikraft.md:18]: Yeah, I didn't see what's to explain, but now I think I should explain why they are similar.

File: content/chapters/compute/lab/README.md
Comment (teodutu) [reply content/chapters/compute/lab/README.md:905]: Yeah, you're right. I changed the use case to something more generic: "more context switches => ULTs are better".

File: content/chapters/compute/lab/support/apache2-simulator/apache2_simulator_condition.py
Comment (teodutu) [reply content/chapters/compute/lab/support/apache2-simulator/apache2_simulator_condition.py:17]: I know, but being Python threads, any form of waiting, either by sleeping or by busy-waiting will cause the same behaviour. Do you see a better approach?

File: content/chapters/compute/lab/quiz/parent-faults-before-fork.md
Comment (Alex-deVis) [inline content/chapters/compute/lab/quiz/parent-faults-before-fork.md:12]:
```suggestion
- `mmap()` sets the pages to be mapped at a later time, decided by the OS

## Feedback

```
Add feedback for this quiz.

File: content/chapters/compute/lab/quiz/child-faults-after-write.md
Comment (Alex-deVis) [inline content/chapters/compute/lab/quiz/child-faults-after-write.md:13]:
```suggestion
- They are caused by the `bash` process forking itself when creating the child process

## Feedback

```
Add feedback for this quiz.

File: content/chapters/compute/lab/quiz/tls-synchronization.md
Comment (Alex-deVis) [inline content/chapters/compute/lab/quiz/tls-synchronization.md:9]: This is incomplete.

**Review by Alex-deVis — CHANGES_REQUESTED** [review]
I left targeted suggestions in this PR[1] to your branch.
In a broader view:
- We should decide on using `synchronisation` or `synchronization` for consistency
- We should consider naming quizzes instead of listing them as "quiz"

[1]: https://github.com/teodutu/operating-systems-oer/pull/3

File: content/chapters/compute/lab/quiz/parent-faults-before-fork.md
Comment (teodutu) [reply content/chapters/compute/lab/quiz/parent-faults-before-fork.md:12]: I wanted to, but I don't know what to write. The feedback is kind of the answer itself.

File: content/chapters/compute/lab/quiz/child-faults-after-write.md
Comment (teodutu) [reply content/chapters/compute/lab/quiz/child-faults-after-write.md:13]: Ditto. The question asks for the cause of the page faults and the answer also provides an explanation.

**Comment by teodutu** [comment]
> We should decide on using synchronisation or synchronization for consistency

Yeah, you're right. I switched to the American spelling :(

> We should consider naming quizzes instead of listing them as "quiz"

So you mean something like `[Quiz - Some Name Here](../quiz/blah-blah.md)`? I don't see a problem with simply writing "Quiz". In the future, they will be interactive so the name of the link is irrelevant.

File: content/chapters/compute/lab/quiz/parent-faults-before-fork.md
Comment (Alex-deVis) [reply content/chapters/compute/lab/quiz/parent-faults-before-fork.md:12]: We could leave a link to a section where demand paging is explained. If that seems ugly, no feedback is also ok.

File: content/chapters/compute/lab/quiz/child-faults-after-write.md
Comment (Alex-deVis) [reply content/chapters/compute/lab/quiz/child-faults-after-write.md:13]: Agreed.

File: content/chapters/compute/lab/quiz/parent-faults-before-fork.md
Comment (teodutu) [reply content/chapters/compute/lab/quiz/parent-faults-before-fork.md:12]: It's ugly and a bit unnecessary. Reading the lab section on demand paging is kind of the default feedback. I'd leave this question without feedback.

**Comment by Alex-deVis** [comment]
> So you mean something like [Quiz - Some Name Here](../quiz/blah-blah.md)?

Yes, I think it makes it easier to navigate the lab.

**Comment by teodutu** [comment]
> > So you mean something like [Quiz - Some Name Here](../quiz/blah-blah.md)?
> 
> Yes, I think it makes it easier to navigate the lab.

The other labs don't follow this convention and I don't see how it improves navigability. You can still click a link to a quiz either way.

**Pull request merged**
